File: windows/src/TokenBar/Services/OpenAIService.cs
using TokenBar.I18n;
using System;
using System.Globalization;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using TokenBar.Helpers;
using TokenBar.Models;

namespace TokenBar.Services
{
    public class OpenAIService
    {
        public static OpenAIService Instance { get; } = new OpenAIService();

        private OpenAIService() { }

        public async Task<(TokenWindow? Primary, TokenWindow? Secondary, string? Account)> FetchQuotaAsync(
            string apiKey,
            string endpoint = "https://api.openai.com/v1",
            string? organizationId = null,
            CancellationToken ct = default)
        {
            var trimmedKey = apiKey.Trim();
            if (string.IsNullOrEmpty(trimmedKey))
            {
                throw new ArgumentException(LocalizationManager.Instance.IsChinese ? "请输入 OpenAI API Key" : "Please enter OpenAI API Key");
            }

            var baseEndpoint = endpoint.Trim().TrimEnd('/');
            if (string.IsNullOrEmpty(baseEndpoint))
            {
                baseEndpoint = "https://api.openai.com/v1";
            }

            var targetUrl = baseEndpoint.EndsWith("/models", StringComparison.OrdinalIgnoreCase)
                ? baseEndpoint
                : $"{baseEndpoint}/models";

            using var request = new HttpRequestMessage(HttpMethod.Get, targetUrl);
            request.Headers.Add("Authorization", $"Bearer {trimmedKey}");
            request.Headers.Add("Accept", "application/json");

            if (!string.IsNullOrWhiteSpace(organizationId))
            {
                request.Headers.Add("OpenAI-Organization", organizationId.Trim());
            }

            HttpResponseMessage response;
            try
            {
                response = await Http.Shared.SendAsync(request, ct);
            }
            catch (OperationCanceledException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new Exception(LocalizationManager.Instance.IsChinese ? $"OpenAI 网络连接失败: {ex.Message}" : $"OpenAI network connection failed: {ex.Message}", ex);
            }

            using (response)
            {
                var responseBody = await response.Content.ReadAsStringAsync(ct);

                if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
                {
                    throw new Exception(LocalizationManager.Instance.IsChinese ? "OpenAI API Key 无效或已过期 (HTTP 401)" : "OpenAI API Key is invalid or expired (HTTP 401)");
                }

                if ((int)response.StatusCode == 429)
                {
                    var msg = LocalizationManager.Instance.IsChinese ? "请求过于频繁或额度已耗尽 (HTTP 429)" : "Rate limit reached or quota exhausted (HTTP 429)";
                    try
                    {
                        using var doc = JsonDocument.Parse(responseBody);
                        if (doc.RootElement.TryGetProperty("error", out var err) &&
                            err.TryGetProperty("message", out var errDetail))
                        {
                            msg = errDetail.GetString() ?? msg;
                        }
                    }
                    catch (JsonException ex)
                    {
                        Log.Warn("provider", $"openai 429 响应不是 JSON: {ex.Message}");
                    }
                    throw new Exception(msg);
                }

                if (!response.IsSuccessStatusCode)
                {
                    var snippet = responseBody.Length > 120 ? responseBody.Substring(0, 120) : responseBody;
                    throw new Exception(LocalizationManager.Instance.IsChinese ? $"OpenAI 接口请求失败 ({(int)response.StatusCode}): {snippet}" : $"OpenAI request failed ({(int)response.StatusCode}): {snippet}");
                }

                // Parse rate limit headers
                string? GetHeader(string name)
                {
                    if (response.Headers.TryGetValues(name, out var values))
                        return values.FirstOrDefault();
                    if (response.Content.Headers.TryGetValues(name, out var contentValues))
                        return contentValues.FirstOrDefault();
                    return null;
                }

                var limitTokensStr = GetHeader("x-ratelimit-limit-tokens");
                var remainingTokensStr = GetHeader("x-ratelimit-remaining-tokens");
                var resetTokensStr = GetHeader("x-ratelimit-reset-tokens");

                var limitReqStr = GetHeader("x-ratelimit-limit-requests");
                var remainingReqStr = GetHeader("x-ratelimit-remaining-requests");
                var resetReqStr = GetHeader("x-ratelimit-reset-requests");

                var orgHeader = GetHeader("openai-organization");

                TokenWindow? primaryWindow = null;
                TokenWindow? secondaryWindow = null;

                // 1. Tokens Rate Limit Window (TPM)
                if (double.TryParse(limitTokensStr, NumberStyles.Float, CultureInfo.InvariantCulture, out var limitTokens) &&
                    double.TryParse(remainingTokensStr, NumberStyles.Float, CultureInfo.InvariantCulture, out var remainingTokens) &&
                    limitTokens > 0)
                {
                    var used = Math.Max(0.0, limitTokens - remainingTokens);
                    var usedPct = Math.Clamp((used / limitTokens) * 100.0, 0.0, 100.0);
                    var resetSec = RateLimitReset.Parse(resetTokensStr) ?? 1;
                    var now = DateTime.Now;
                    var resetDate = now.AddSeconds(resetSec);

                    primaryWindow = new TokenWindow
                    {
                        Title = "TPM 速率剩余",
                        UsedPercentage = usedPct,
                        StartTime = now,
                        EndTime = resetDate,
                        UsedAmount = used,
                        TotalLimit = limitTokens,
                        Unit = "tokens",
                        IsIdle = used == 0.0
                    };
                }

                // 2. Requests Rate Limit Window (RPM)
                if (double.TryParse(limitReqStr, NumberStyles.Float, CultureInfo.InvariantCulture, out var limitReq) &&
                    double.TryParse(remainingReqStr, NumberStyles.Float, CultureInfo.InvariantCulture, out var remainingReq) &&
                    limitReq > 0)
                {
                    var used = Math.Max(0.0, limitReq - remainingReq);
                    var usedPct = Math.Clamp((used / limitReq) * 100.0, 0.0, 100.0);
                    var resetSec = RateLimitReset.Parse(resetReqStr) ?? 1;
                    var now = DateTime.Now;
                    var resetDate = now.AddSeconds(resetSec);

                    secondaryWindow = new TokenWindow
                    {
                        Title = "RPM 请求速率",
                        UsedPercentage = usedPct,
                        StartTime = now,
                        EndTime = resetDate,
                        UsedAmount = used,
                        TotalLimit = limitReq,
                        Unit = "req",
                        IsIdle = used == 0.0
                    };
                }

                int modelCount = 0;
                try
                {
                    using var doc = JsonDocument.Parse(responseBody);
                    if (doc.RootElement.TryGetProperty("data", out var dataArr) &&
                        dataArr.ValueKind == JsonValueKind.Array)
                    {
                        modelCount = dataArr.GetArrayLength();
                    }
                }
                catch (JsonException ex)
                {
                    Log.Warn("provider", $"openai /models 响应不是 JSON: {ex.Message}");
                }

                if (primaryWindow == null)
                {
                    // 没有速率头：只表示连接正常，不编造任何百分比
                    primaryWindow = TokenWindow.Status("API 连接状态");
                }

                var accountInfo = orgHeader ?? organizationId;
                if (string.IsNullOrWhiteSpace(accountInfo))
                {
                    accountInfo = modelCount > 0
                        ? (LocalizationManager.Instance.IsChinese ? $"OpenAI (可用模型: {modelCount}个)" : $"OpenAI ({modelCount} models available)")
                        : "OpenAI API";
                }

                return (primaryWindow, secondaryWindow, accountInfo);
            }
        }
    }
}

File: mac/Sources/TokenBar/Services/OpenAIService.swift
import Foundation

public final class OpenAIService: @unchecked Sendable {
    public static let shared = OpenAIService()

    public init() {}

    /// Fetches OpenAI quota and rate limits
    public func fetchQuota(
        apiKey: String,
        endpoint: String = "https://api.openai.com/v1",
        organizationId: String? = nil
    ) async throws -> (primary: TokenWindow?, secondary: TokenWindow?, account: String?) {
        let trimmedKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedKey.isEmpty else {
            let isZh = LocalizationManager.shared.effectiveLanguage == "zh"
            throw NSError(
                domain: "OpenAIService",
                code: 400,
                userInfo: [NSLocalizedDescriptionKey: isZh ? "请输入 OpenAI API Key" : "Please enter OpenAI API Key"]
            )
        }

        var baseEndpoint = endpoint.trimmingCharacters(in: .whitespacesAndNewlines)
        if baseEndpoint.isEmpty {
            baseEndpoint = "https://api.openai.com/v1"
        }
        while baseEndpoint.hasSuffix("/") {
            baseEndpoint.removeLast()
        }

        let targetURLString: String
        if baseEndpoint.hasSuffix("/models") {
            targetURLString = baseEndpoint
        } else {
            targetURLString = "\(baseEndpoint)/models"
        }

        guard let url = URL(string: targetURLString) else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(trimmedKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let org = organizationId?.trimmingCharacters(in: .whitespacesAndNewlines), !org.isEmpty {
            request.setValue(org, forHTTPHeaderField: "OpenAI-Organization")
        }
        request.timeoutInterval = 12

        let (data, response) = try await HTTPClient.data(for: request)
        guard let httpResp = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }

        let isZh = LocalizationManager.shared.effectiveLanguage == "zh"
        if httpResp.statusCode == 401 {
            throw NSError(
                domain: "OpenAIService",
                code: 401,
                userInfo: [NSLocalizedDescriptionKey: isZh ? "OpenAI API Key 无效或已过期 (HTTP 401)" : "OpenAI API Key is invalid or expired (HTTP 401)"]
            )
        }

        if httpResp.statusCode == 429 {
            var msg = isZh ? "请求过于频繁或额度已耗尽 (HTTP 429)" : "Rate limit reached or quota exhausted (HTTP 429)"
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let err = json["error"] as? [String: Any],
               let errDetail = err["message"] as? String {
                msg = errDetail
            }
            throw NSError(domain: "OpenAIService", code: 429, userInfo: [NSLocalizedDescriptionKey: msg])
        }

        guard httpResp.statusCode >= 200 && httpResp.statusCode < 300 else {
            let errorMsg = String(data: data, encoding: .utf8) ?? "HTTP \(httpResp.statusCode)"
            throw NSError(
                domain: "OpenAIService",
                code: httpResp.statusCode,
                userInfo: [NSLocalizedDescriptionKey: isZh ? "OpenAI 接口请求失败 (\(httpResp.statusCode)): \(errorMsg.prefix(120))" : "OpenAI request failed (\(httpResp.statusCode)): \(errorMsg.prefix(120))"]
            )
        }

        // Parse rate limits from headers
        let allHeaders = httpResp.allHeaderFields

        func getHeader(_ name: String) -> String? {
            let lowerTarget = name.lowercased()
            for (k, v) in allHeaders {
                if let keyStr = k as? String, keyStr.lowercased() == lowerTarget {
                    return String(describing: v)
                }
            }
            return nil
        }

        let limitTokensStr = getHeader("x-ratelimit-limit-tokens")
        let remainingTokensStr = getHeader("x-ratelimit-remaining-tokens")
        let resetTokensStr = getHeader("x-ratelimit-reset-tokens")

        let limitReqStr = getHeader("x-ratelimit-limit-requests")
        let remainingReqStr = getHeader("x-ratelimit-remaining-requests")
        let resetReqStr = getHeader("x-ratelimit-reset-requests")

        let orgHeader = getHeader("openai-organization")

        var primaryWindow: TokenWindow? = nil
        var secondaryWindow: TokenWindow? = nil

        // 1. Tokens Rate Limit Window (TPM)
        if let limitTokens = Double(limitTokensStr ?? ""),
           let remainingTokens = Double(remainingTokensStr ?? ""),
           limitTokens > 0 {
            let used = max(0.0, limitTokens - remainingTokens)
            let usedPct = min(max((used / limitTokens) * 100.0, 0.0), 100.0)

            let duration = parseDurationString(resetTokensStr ?? "1s")
            let now = Date()
            let resetDate = now.addingTimeInterval(duration)

            primaryWindow = TokenWindow(
                title: "TPM 速率剩余",
                usedPercentage: usedPct,
                startTime: now,
                endTime: resetDate,
                usedAmount: used,
                totalLimit: limitTokens,
                unit: "tokens",
                isIdle: used == 0.0
            )
        }

        // 2. Requests Rate Limit Window (RPM)
        if let limitReq = Double(limitReqStr ?? ""),
           let remainingReq = Double(remainingReqStr ?? ""),
           limitReq > 0 {
            let used = max(0.0, limitReq - remainingReq)
            let usedPct = min(max((used / limitReq) * 100.0, 0.0), 100.0)

            let duration = parseDurationString(resetReqStr ?? "1s")
            let now = Date()
            let resetDate = now.addingTimeInterval(duration)

            secondaryWindow = TokenWindow(
                title: "RPM 请求速率",
                usedPercentage: usedPct,
                startTime: now,
                endTime: resetDate,
                usedAmount: used,
                totalLimit: limitReq,
                unit: "req",
                isIdle: used == 0.0
            )
        }

        // If no rate limit headers are exposed by the gateway, show connected status
        var modelCount = 0
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let dataArr = json["data"] as? [[String: Any]] {
            modelCount = dataArr.count
        }

        if primaryWindow == nil {
            primaryWindow = TokenWindow.status(title: "API 连接状态")
        }

        var accountInfo = orgHeader ?? organizationId
        if accountInfo == nil || accountInfo?.isEmpty == true {
            accountInfo = modelCount > 0 ? (isZh ? "OpenAI (可用模型: \(modelCount)个)" : "OpenAI (\(modelCount) models available)") : "OpenAI API"
        }

        return (primaryWindow, secondaryWindow, accountInfo)
    }

    /// 兼容旧调用：委托给 `RateLimitReset.parse`，无法解析时按 1 秒兜底，
    /// 最小 0.1 秒（毫秒级重置对倒计时没有意义，沿用旧下限）。
    public func parseDurationString(_ str: String) -> TimeInterval {
        max(0.1, RateLimitReset.parse(str) ?? 1.0)
    }
}
